core/chains/evm/gas/universal_estimator.go

```diff
@@ -0,0 +1,412 @@
+	for {
+		select {
+		case <-u.stopCh:
+			return
+		case <-t.C:
+			if u.config.EIP1559 {
+				if _, err := u.FetchDynamicPrice(); err != nil {
+					u.logger.Error(err)
+				}
+			} else {
+				if _, err := u.FetchGasPrice(); err != nil {
+					u.logger.Error(err)
+				}
+			}
+		}
+	}
+}
+
+// GetLegacyGas will fetch the cached gas price value.
+func (u *UniversalEstimator) GetLegacyGas(ctx context.Context, _ []byte, gasLimit uint64, maxPrice *assets.Wei, opts ...feetypes.Opt) (gasPrice *assets.Wei, chainSpecificGasLimit uint64, err error) {
+	chainSpecificGasLimit = gasLimit
+	if gasPrice, err = u.getGasPrice(); err != nil {
+		return
+	}
+
+	if gasPrice.Cmp(maxPrice) > 0 {
+		u.logger.Warnf("estimated gas price: %s is greater than the maximum gas price configured: %s, returning the maximum price instead.", gasPrice, maxPrice)
+		return maxPrice, chainSpecificGasLimit, nil
+	}
+	return
+}
+
+// FetchGasPrice will use eth_gasPrice to fetch and cache the latest gas price from the RPC.
+func (u *UniversalEstimator) FetchGasPrice() (*assets.Wei, error) {
+	ctx, cancel := u.stopCh.CtxCancel(evmclient.ContextWithDefaultTimeout())
+	defer cancel()
+
+	gasPrice, err := u.client.SuggestGasPrice(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch gas price: %s", err)
+	}
+
+	promUniversalEstimatorGasPrice.WithLabelValues(u.chainID.String()).Set(float64(gasPrice.Int64()))
+
+	gasPriceWei := (*assets.Wei)(gasPrice)
+
+	u.logger.Debugf("fetched new gas price: %v", gasPriceWei)
+
+	u.gasPriceMu.Lock()
+	defer u.gasPriceMu.Unlock()
+	u.gasPrice = gasPriceWei
+	return u.gasPrice, nil
+}
+
+func (u *UniversalEstimator) getGasPrice() (*assets.Wei, error) {
+	u.gasPriceMu.RLock()
+	defer u.gasPriceMu.RUnlock()
+	if u.gasPrice == nil {
+		return u.gasPrice, fmt.Errorf("gas price not set")
+	}
+	return u.gasPrice, nil
+}
+
+// GetDynamicFee will fetch the cached dynamic prices.
+func (u *UniversalEstimator) GetDynamicFee(ctx context.Context, maxPrice *assets.Wei) (fee DynamicFee, err error) {
+	if fee, err = u.getDynamicPrice(); err != nil {
+		return
+	}
+
+	if fee.FeeCap == nil || fee.TipCap == nil {
+		return fee, fmt.Errorf("dynamic price not set")
+	}
+	if fee.FeeCap.Cmp(maxPrice) > 0 {
+		u.logger.Warnf("estimated maxFeePerGas: %v is greater than the maximum price configured: %v, returning the maximum price instead.",
+			fee.FeeCap, maxPrice)
+		fee.FeeCap = maxPrice
+		if fee.TipCap.Cmp(maxPrice) > 0 {
+			u.logger.Warnf("estimated maxPriorityFeePerGas: %v is greater than the maximum price configured: %v, returning the maximum price instead.",
+				fee.TipCap, maxPrice)
+			fee.TipCap = maxPrice
+		}
+	}
+
+	return
+}
+
+// FetchDynamicPrice uses eth_feeHistory to fetch the basFee of the latest block and the Nth maxPriorityFeePerGas percentiles
+// of the past X blocks. It also fetches the highest Zth maxPriorityFeePerGas percentile of the past X blocks. Z is configurable
+// and it represents the highest percentile we're willing to pay.
```

Comment:
Since `ConnectivityPercentile` is set as a const above, I think this is technically not true anymore.
```suggestion
// of the past X blocks. It also fetches the highest Zth maxPriorityFeePerGas percentile of the past X blocks.
```

Reply:
I wanted to point out that it can be changed, but I'll just add the fixed value instead.

Reply:
I think the comment is still a little outdated. 
> It also fetches the highest 85th maxPriorityFeePerGas percentile of the past X blocks. Z is configurable and it represents the highest percentile we're willing to pay.

should be

> It also fetches the highest 85th maxPriorityFeePerGas percentile of the past X blocks which represents the highest percentile we're willing to pay.